[Foundation] Modernize hashing in Foundation's Swift-only types

lorentey opened the pull request:
This PR upgrades the Foundation overlay to use the new hashing API introduced in [SE-0206]. This gets rid of the deprecation warnings for `hashValue` implementations and brings the overlay in sync with current Swift best practices.

[SE-0206]: https://github.com/apple/swift-evolution/blob/master/proposals/0206-hashable-enhancements.md

## Background 

[SE-0206] radically changed how hashing works in Swift 4.2+. It is now the stdlib's responsibility to choose a hash function; types implementing `Hashable` only need to worry about selecting what they want to feed to the hasher. To make this possible, `Hashable` now requires hashing to be implemented through the `hash(into:)` method. 

The old `hashValue` property is deprecated as a `Hashable` requirement, and it's only kept to support source-level compatibility with existing code. `Set` and `Dictionary` have switched over to using `hash(into:)` as their entry point for hashing; they do not directly call `hashValue` at all.

Beyond simplifying hashing, the intent of SE-0206 is to enable Swift to provide certain guarantees about its quality. In particular: as long as `hash(into:)` feeds enough data the hasher to unambiguously decide equality, *Swift attempts to guarantee that collision attacks won't be possible*. For this to work, it is critically important for `Hashable` implementations to include everything that `Equatable.==` looks at; and this is *especially* the case for the basic boundary types that come built-in with Swift, like `Data`.

The exact hash algorithm used is a private implementation detail of the stdlib; the `Hasher` API was carefully designed to leak as little information about it as possible, enabling future versions of the stdlib to switch to other algorithms without worrying about ABI compatibility issues.

While random seeding makes the hash algorithm opaque, it is technically possible for Swift code to rely on the particular hash encoding of a stdlib- (or SDK-)provided type. This is highly unlikely to happen in practice, but the longer we wait to fix hashing issues, the more likely it will be that someone starts relying on hashing particulars. For example, I could imagine this to happen through a well-meaning desire to work around a broken hash implementation. It is therefore important to do this sooner rather than later.

## Notes

I'll highlight points specific to a particular changes as self-review comments. Here is a list of general remarks:

1. It is neither required nor desirable for Swift's `hashValue` to return reproducible values.

2. We don't need to ever match hash values between Swift types and the Cocoa classes that they bridge with. The act of bridging involves a full rehashing of all Set/Dictionary values.

3. The only hard requirement is that hashing needs to be consistent with the equivalence classes implemented by `==` *within the same type*.

4. Hashing is meaningless without equality. `Equatable` only defines equality for values of a single type. Therefore, `Hashable` makes no requirements or expectations about the hash values produced by values of different types. 

    - It is perfectly fine for hash values produced by different types to not match. `(0 as Int8).hashValue` and `(0 as Int16).hashValue` will (typically) not be equal in Swift.

    - It is also perfectly fine for values of two distinct types to produce matching hash values. `Measurement<UnitLength>.hash(into:)` and `Measurement<UnitDuration>.hash(into:)` are allowed to use the same hash encoding, since there is no way they can ever produce large-scale collisions within the same hash table.

5. While this is not a hard requirement for user code, for boundary types provided in the stdlib/SDK, we require that hashing isn't just consistent with equality, but that it's *equivalent* to it. The Swift test suite has [checks] to actively enforce this -- this is possible through [repeatedly salting the hash function][salt].

    "Optimizing" hashing by omitting some of the data compared by `==` is generally a mistake in Swift, because it *completely* breaks all guarantees about the strength of hashing, and opens the door to (accidental or deliberate) collision attacks.

    It's perfectly acceptable to hash a gigabyte of data if someone inserts some large value (such as a big collection) as a key in a hash table. Multi-megabyte String keys are easy to protect against; hidden hashing weaknesses aren't.

[checks]: https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/StdlibUnittest.swift#L2417-L2578
[salt]: https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/StdlibUnittest.swift#L2541-L2562

6. It would be nice if we could somehow improve hashing in Foundation's Objective-C classes, but this is outside the scope of this PR (or indeed, this repository). However, we should at least make sure Swift programmers can rely on the quality of hashes produced by native Swift value types.

7. https://github.com/apple/swift-corelibs-foundation/pull/2118 is the corresponding PR for swift-corelibs-foundation. These PRs should land at the same time.

8. Replacing `hashValue` implementations on types with `hash(into:)` has no effect on the symbols exposed as ABI -- the compiler automatically synthesizes `hashValue` from `hash(into:)`, and vice versa. (This is not the case in protocol extensions, but we've fixed those in the 4.2/5.0 timeframe.) In the Swift 5.0 ABI, all of these types include definitions for both `hashValue` and `hash(into:) so we don't need to worry about availability issues.

    ~~The sole exception is `URL`, which implements `hashValue` but isn't `Hashable`. Unfortunately we can't add the conformance in 5.1, but we should add a (versioned) `hash(into:)` definition to discourage people from trying to implement it on their own.~~ (Edit: `URL` is not a special case; it's `Hashable` like the rest. I got mislead by an unrelated error.)


rdar://problem/43394032

lorentey committed:
[test] StdlibUnittest: Cosmetic changes to checkEquatable/checkHashable

lorentey committed:
[Foundation] IndexPath: Add explicit hash(into:) definition, hashing *all* components

The new definition diverges from the existing hashValue: it feeds all path components to the hasher, rather than just a limited subset.

Hashing all the bits that are compared in == is necessary to ensure proper operation of the hash table.

lorentey committed:
[Foundation] NSRange: Add an explicit definition for hash(into:)

This replaces the compiler-generated variant (based on hashValue) with a better implementation.

lorentey committed:
[Foundation] Calendar: Modernize hashing

lorentey committed:
[Foundation] CharacterSet: Modernize hashing

lorentey committed:
[StdlibUnittest] checkHashable: Add opt-in support for incomplete hashes

lorentey committed:
[Foundation] String.Encoding: Modernize hashing

lorentey committed:
[Foundation] AffineTransform: modernize hashing

lorentey committed:
[Foundation] Data: Hash the entire contents, not just an arbitrary subset

lorentey committed:
[Foundation] Date: Modernize hashing

lorentey committed:
[Foundation] DateComponents: Modernize hashing

lorentey committed:
[Foundation] DateInterval: Modernize hashing

lorentey committed:
[Foundation] Decimal: Modernize hashing

lorentey committed:
[Foundation] Locale: Modernize hashing

lorentey committed:
[Foundation] IndexSet: Modernize hashing

lorentey committed:
[Foundation] IndexPath: Modernize hashing

lorentey committed:
[Foundation] Measurement: Fix hashing

lorentey committed:
[Foundation] Notification: Modernize hashing

lorentey committed:
[Foundation] NSRange: Modernize hashing

lorentey committed:
[Foundation] TimeZone: Modernize hashing

lorentey committed:
[Foundation] URL: Add hash(into:) implementation

Note: URL does not implement Hashable.

lorentey committed:
[Foundation] URLComponents: Modernize hashing

lorentey committed:
[Foundation] URLRequest: Modernize hashing

lorentey committed:
[Foundation] UUID: Modernize hashing

lorentey committed:
[Foundation] PersonNameComponents: Modernize hashing

phausler commented:
@itaiferber @millenomi can y’all give this a run through? There are a number of changes here but the general concept seems well founded. I am worried that I am missing any hidden gotchas to binary comparability or other developer ergonomics etc.

lorentey committed:
[Foundation] URL: Fix availability of the new hash(into:) implementation

itaiferber commented on stdlib/public/Darwin/Foundation/Calendar.swift:
```diff
@@ -899,15 +899,16 @@ public struct Calendar : Hashable, Equatable, ReferenceConvertible, _MutableBoxi
+    public func hash(into hasher: inout Hasher) {
+        // We need to make sure autoupdating calendars have the same hash
+	      if _autoupdating {
+            hasher.combine(false)
```
Not sure if GitHub or there's inconsistent indentation here, but it would be nice to be consistent.

lorentey replied:
It must be a tabs vs spaces thing; I'll take a look.

millenomi commented:
My main concern, which is just top-of-my-head this very moment, is: what about AnyHashable equality? While all the points about Hashable hold, IIRC currently we hold `AnyHashable(x) == AnyHashable(x as AnyObject)` for any object that implements `_ObjectiveCBridgeable`.

itaiferber commented on stdlib/public/Darwin/Foundation/Data.swift:
```diff
@@ -1046,9 +1046,8 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
-            let range = startIndex ..< Swift.min(startIndex + 80, endIndex)
-            storage.withUnsafeBytes(in: range) {
+            // To ensure strong hashing, all bytes must be fed into the hasher.
+            withUnsafeBytes {
```
I'm still not convinced about this without `Hasher` having some guarantees about not being `O(n)`, but I think we should try to set up a meeting to discuss about strategy here.

lorentey replied:
OK! I've spun off this change into #23876. `Data` is already implementing the right APIs, so this is just a matter of changing the implementation.

millenomi commented:
I want to highlight the IIRC; I'm not immediately sure and I have to research this, but I do know the ergonomics can be complex there esp. wrt newly bridged dictionaries of `[AnyHashable: Any]` type.

itaiferber commented on stdlib/public/Darwin/Foundation/IndexPath.swift:
```diff
@@ -662,25 +662,31 @@ public struct IndexPath : ReferenceConvertible, Equatable, Hashable, MutableColl
-            return count &+ (first << lengthBits) &+ (last << (lengthBits + firstIndexBits))
-        }
+    public func hash(into hasher: inout Hasher) {
+	      // Note: We compare all indices in ==, so for proper hashing, we must
```
Also, weird indentation here?

itaiferber commented:
Overall, most of the changes here look reasonable (and very welcome!) to me.

parkera commented on stdlib/public/Darwin/Foundation/URL.swift:
```diff
@@ -622,10 +622,14 @@ public struct URL : ReferenceConvertible, Equatable {
     }
     
-    public var hashValue: Int {
+    public var hashValue: Int { // FIXME(hashValue): This should be removed once URL implements Hashable.
```
What’s the limitation here, and how temporary is it?

lorentey replied:
There is currently no way to mark the availability of protocol conformances, or to add a conformance in a way that deploys back to the Swift 5 ABI (rdar://problem/35158274).

This means that new conformances to protocols that shipped in Xcode 10.2 cannot be added to types that also exist in same.

lorentey replied:
Update: I was actually wrong -- `URL` conforms to `Hashable` by way of `ReferenceConvertible`; the conformance is just not explicitly declared. 

I must have confused an unrelated error with a missing conformance when I double checked this. 🤦🏻‍♂️

This is a huge relief: we can simply replace `hashValue` with `hash(into:)` without any availability tricks.

lorentey commented:
@millenomi The bridgeable types I touched should all implement `_toCustomAnyHashable()` by bridging to Swift, so AnyHashable behavior wouldn't regress by these changes. 

Possible weird edge case: I can't figure out where `Decimal`/`NSDecimalNumber` get their `_HasCustomAnyHashableRepresentation` conformances. They seem to do the right thing, though.

lorentey committed:
[Foundation] URL is actually already Hashable

I’m not sure why I was so convinced it wasn’t, but this certainly simplifies things.

lorentey committed:
[Foundation] Fix bizarre indentation issues

lorentey committed:
Revert "[Foundation] Data: Hash the entire contents, not just an arbitrary subset"

This reverts commit b711ed9f90222302cebc6ae6e36ba7eb755a7d83.

lorentey committed:
[Foundation] Notification: Add note on == not being reflexive and stabilize hashing

The `ObjectIdentifier(object as AnyObject)` is not necessarily stable; this breaks reflexivity for ==, and it makes the hash encoding nondeterministic.

lorentey commented:
I submitted a PR for swift-corelibs-foundation with matching changes: https://github.com/apple/swift-corelibs-foundation/pull/2118

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test macOS platform

lorentey committed:
[test] TestNotification: Fix checkHashable invocation

lorentey commented:
@swift-ci please test macOS platform

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test

lorentey commented:
```
02:13:44 ******************** TEST 'Swift(iphonesimulator-i386) :: stdlib/TestPersonNameComponents.swift' FAILED ********************
...
02:13:44 [ RUN      ] TestPersonNameComponents.test_Hashing
02:13:44 stdout>>> check failed at /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/stdlib/TestPersonNameComponents.swift, line 54
02:13:44 stdout>>> stacktrace:
02:13:44 stdout>>>   #0: /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/stdlib/private/StdlibUnittest/StdlibUnittest.swift:2530
02:13:44 stdout>>> expected: 230665695 (of type Swift.Int)
02:13:44 stdout>>> actual: -422011617 (of type Swift.Int)
02:13:44 stdout>>> hash(into:) expected to match, found to differ
02:13:44 stdout>>> lhs (at index 0): givenName: Kevin familyName: Frank 
02:13:44 stdout>>> rhs (at index 1): givenName: Kevin familyName: Frank 
```

Hm; `NSPersonNameComponents.hash` seems not to work correctly on iphonesimulator-i386.

lorentey committed:
[test] TestPersonNameComponents: adjust OS check for hashing

lorentey commented:
@swift-ci please test

airspeedswift commented:
@parkera @millenomi @itaiferber are we good to go ahead with this (with the `Data` part subsetted out for further discussion)

parkera commented:
Yup

itaiferber reviewed (approved):
Latest changes LGTM
